docs/features/source-generators.cookbook.md

```diff
@@ -0,0 +1,250 @@
+            foreach (var file in myFiles)
+            {
+                var content = file.GetText(context.CancellationToken);
+
+                // do some transforms based on the file contenxt
+
+                string output = "namespace MyNS { public class MyClass { } }";
+                var sourceText = SourceText.From(output);
+
+                context.AddSource($"{file.Name}generated.cs", sourceText);
+            }
+        }
+    }
+```
+
+### Augment user code
+
+**User scenario:** As a generator author I want to be able to augment a users code with new functionality.
+
+**Solution:** Require the user to make the class you want to augment be a `partial class`, and mark it with e.g. a unique attribute, or name.
+Look for any classes marked for generation and generate the a matching `partial class` that contains the additional functionality.
+
+**Example:**
+
+```csharp
+public partial class UserClass
+{
+    public void UserMethod()
+    {
+        // call into a method inside the class
+        this.GeneratedMethod();
+    }
+}
+```
+
+```csharp
+public class AugmentingGenerator : ISourceGenerator
+    {
+        public void Execute(SourceGeneratorContext context)
+        {
+            var syntaxTrees = context.Compilation.SyntaxTrees;
+            foreach (var syntaxTree in syntaxTrees)
+            {
+                 // find the class to augment
+                var classToAugment = syntaxTree.GetRoot().DescendantNodes()
+                 .OfType<ClassDeclarationSyntax>()
+                 .Where(c => c.Name.ToString() == "UserClass")
+                 .Single();
+
+                var sourceText = SourceText.From($@"
+public partial class {classToAugment.Identifier.ToString()}
+{
+   private void GeneratedMethod()
+   {
+       // generated code
+   }
+}");
+                context.AddSource("myGeneratedFile.cs", sourceText);
+            }
+        }
+    }
+
+```
+
+### Participate in the IDE experience
+
+**User scenario:** As a generator author I want to be able to interactively regenerate code as the user is editing files.
+
+**Solution:** We expect there to be an an opt-in set of interactive interfaces that can be implemented to allow for progressively more complex generation strategies.
+It is anticipated there will be a mechanism for providing symbol mapping for lighting up features such a 'Find all references'.
+
+```csharp
+public class InteractiveGenerator : ISourceGenerator, IAdditionalFilesChangedGenerator
+    {
+        public void Execute(SourceGeneratorContext context)
+        {
+            // generators must always support a total generation pass
+        }
+
+        public void OnAdditionalFilesChanged(AdditionalFilesChangedContext context)
+        {
+            // determine which file changed, and if it affects this generator
+            // regenerate only the parts that are affected by this change.
+        }
+    }
+```
+
+### INotifyPropertyChanged
```

Comment:
We should probably mention what our expectation is for providing this attributed. Would this be duck typing? IOW, would the attribute live in user code as an internal type? If so, we should probably mention that the generator should provide a code fix that can be keyed of, for example, `INotifyPropertyChanged` on the class direction.